### ISSUE TITLE
MongoDB URI database handling not aligned with MongoDB URI specification

### ISSUE THREAD

**Issue by sourcekick** [scullxbones/akka-persistence-mongo#75]
**MongoDB URI specification**

In a MongoDB URI, the optional database path element is ONLY there to optionally define the authentication database, i.e. the database against which optional user authentication is performed. If no database is given in a MongoDB URI, it needs to default to `admin` as the authentication database. Whether authentication actually has to be performed is determined by the MongoDB server.

The database to perform the actual read and writes is, according to the MongoDB specification, never given in the MongoDB URI. In the following I refer to it as the "effective" database.

See https://docs.mongodb.org/v3.0/reference/connection-string/ and for an example implementation that actually behaves like that (just to have some understatement) see https://github.com/mongodb/mongo-java-driver/releases and http://projects.spring.io/spring-data-mongodb/ .

**Problem**

The following code fragement in akka.contrib.persistence.mongodb.CasbahPersistenceExtension is currently wrong, because it takes the effective database from the MongoDB URI and otherwise defaults to `akka-persistence`:
```
  private[mongodb] lazy val db = client(databaseName.getOrElse(url.database.getOrElse(DEFAULT_DB_NAME)))
```

**Solution Considerations**

1. A possible solution could be to first take the name of the effective database from the existing configuration property `akka.contrib.persistence.mongodb.mongo.db`. Danger: This could break some projects.

2. Another solution could be to introduce a new property for the name of the effective database and use it first, like for example `akka.contrib.persistence.mongodb.mongo.database`. This would only break projects that rely already on the "wrong" semantics of the currently implemented MongoDB URI intepretation.

**Comment by scullxbones** [comment]
Hi @sourcekick,

Can you take a look at [this PR](https://github.com/scullxbones/akka-persistence-mongo/pull/67) that was recently contributed?

I believe it fully addresses this issue via a configuration parameter `database`.  The documentation should be up to date as well, but if it's inaccurate or confusing please let me know.

**Comment by sourcekick** [comment]
Even though I am following mails regarding your plugin I somehow missed that, because we are still using the 0.x Akka 2.3 stream.

The PR is a valid fix in my opinion. Looking forward to a release of the 1.x Akka 2.4 stream including this fix.

It would be nice to also have it fixed in the 0.x Akka 2.3 stream.

Regardind the documentation: consider to update the documentation slightly to make it clear that the plugin does not (or in most cases not?) honor a Write Concern defined in the MongoDB URI, because the Write Concern is explicitely set on each operation when using the MongoDB driver API. At least for Casbah it looks like that. (I think there my be very good reasons for how it is implemented and configurable.)

**Comment by scullxbones** [comment]
This fix is available as of `1.0.11` of the 2.4.x plugin.  It has not been back-ported to the `0.x` series however.

I can add that information to the documentation, no problem.  You're right, write concern is applied from the configuration, overriding anything in the URL.

**Comment by scullxbones** [comment]
Closing this as addressed, no plan currently to backport.
